style_guide.rst

```diff
@@ -0,0 +1,393 @@
+descriptive the name needs to be.
+
+
+..
+    Printing logging / debug messages
+    ---------------------------------
+
+    All the projects in this class use a simple logging library called ``chilog`` that is documented in each of the project specifications. You must use the ``chilog`` functions *exclusively* for printing logging or debug messages. **Do not use printf() directly in your code**. Please note that the ``chilog`` functions provide essentially the same functionality as ``printf``, so there is no situation where ``printf`` would be necessary instead of ``chilog`` (using ``chilog`` consistently also means you will not have to scrub ``printf``'s from your code before submitting it).
+
+    Furthermore, all the messages at the ``INFO``, ``WARNING``, ``ERROR``, and ``CRITICAL`` levels must be used only for their intended purposes (e.g., only use ``ERROR`` to print out actual errors in the execution of your program). You must use the ``DEBUG`` level only to print informative debug messages that would be understood by any developer trying to debug your code. You may use the ``TRACE`` level to print *any* debug message (including those that would only be understood by you). However, if your code is riddled with ``TRACE`` logging statements (including commented out ones) to the point where it is hard to read the code itself, we may take points off for this. So, once a ``TRACE`` logging statement has served its purpose, we suggest you remove it (not just comment it out).
+
+    You should assume that graders will run your code with logging at the ``INFO`` level, and will only use the ``DEBUG`` level if they need to debug an issue with your code. We will never run your code with logging at the ``TRACE`` level.
+
+    Please note that, in assignments where you are responsible for writing the ``main`` function, you may use ``fprintf`` to print to *standard error* if there is an error that prevents the program from starting (e.g., if a command-line parameter has not been provided, etc.)
+
+Global variables
+----------------
+
+The use of global variables is forbidden, except for defining compile-time *constants*. **There are no other exceptions**.
+
+A compile-time constant is a global variable with a value that is known and set at compile-time and *never* changed during the runtime of the program (if the variable can be used to access other values, e.g., because it is a pointer, a ``struct``, etc., all the values reachable from the variable must also be known and set at compile-time). Such variables must have ``ALL_CAPS`` names, be declared as ``const`` and, if they are only going to be used in a single module, they must also be declared as ``static``. Please note that our definition excludes variables that are initialized when the program starts running, *even if the value of that variable won't change once it has been initialized*. If the value of the variable is not known before the program is running, it cannot be a global variable.
+
```

Comment:
Maybe we could add that, if you find yourself wanting to use global variables to keep track of some kind of state, you should consider using a context struct. (But maybe this is too specific for this guide)

Reply:
I think that might be too specific for this guide. However, I can imagine the developer guide having a "best practices" section in the future that explains things like context obejcts (and then we could just have this part of the style guide link to it)